net/usbwlan.c

```diff
@@ -0,0 +1,656 @@
+		}
+
+		dev->pipeBulkIn = usb_open(drv, ins, usb_transfer_bulk, usb_dir_in);
+		if (dev->pipeBulkIn < 0) {
+			usbwlan_log("Fail to open bulk IN pipe");
+			err = -EINVAL;
+			break;
+		}
+
+		dev->pipeBulkOut = usb_open(drv, ins, usb_transfer_bulk, usb_dir_out);
+		if (dev->pipeBulkOut < 0) {
+			usbwlan_log("Fail to open bulk OUT pipe");
+			err = -EINVAL;
+			break;
+		}
+
+		dev->rx.urbid = usb_urbAlloc(dev->drv, dev->pipeBulkIn, NULL, usb_dir_in, USBWLAN_BULK_SIZE, usb_transfer_bulk);
+		if (dev->rx.urbid < 0) {
+			usbwlan_log("Fail to allocate URB");
+			err = -ENOMEM;
+			break;
+		}
+
+		oid.port = usbwlan_common.msgport;
+		oid.id = idtree_id(&dev->node);
+		snprintf(dev->path, sizeof(dev->path), "/dev/wlan%u", oid.id);
+		if (create_dev(&oid, dev->path) < 0) {
+			usbwlan_log("Fail to create device file");
+			err = -EINVAL;
+			break;
+		}
+	} while (0);
+
+	if (err < 0) {
+		_usbwlan_put(dev);
+	}
+
+	mutexUnlock(usbwlan_common.lock);
+
+	if (err < 0) {
+		usbwlan_free(dev);
+		return err;
+	}
+
+	event_out->deviceCreated = true;
+	event_out->dev = oid;
+	(void)snprintf(event_out->devPath, sizeof(event_out->devPath), "%s", dev->path);
+
+	usb_setConfiguration(dev->drv, dev->pipeCtrl, 1);
+
+	return 0;
+}
+
+
+static int usbwlan_handleDeletion(usb_driver_t *drv, usb_deletion_t *del)
+{
+	rbnode_t *node, *next;
+	usbwlan_dev_t *dev;
+
+	mutexLock(usbwlan_common.lock);
+
+	node = lib_rbMinimum(usbwlan_common.devices.root);
+	while (node != NULL) {
+		next = lib_rbNext(node);
+		dev = lib_treeof(usbwlan_dev_t, node, lib_treeof(idnode_t, linkage, node));
+
+		assert(dev != NULL);
+		if (dev->dev.bus == del->bus && dev->dev.dev == del->dev && dev->dev.interface == del->interface) {
+			if (_usbwlan_put(dev) == 0) {
+				usbwlan_free(dev);
+			}
+		}
+
+		node = next;
+	}
+
+	mutexUnlock(usbwlan_common.lock);
+
+	return 0;
+}
+
+
+static int usbwlan_handleCompletion(struct usb_driver *drv, usb_completion_t *completion, const char *data, size_t len)
+{
+	int err;
+	usbwlan_dev_t *dev;
+
+	dev = usbwlan_getByPipe(completion->pipeid);
+	if (dev == NULL) {
+		return -1;
+	}
+
+	if (completion->pipeid != dev->pipeBulkIn) {
+		usbwlan_put(dev);
+		return -1;
+	}
+
+	mutexLock(dev->rx.lock);
+	do {
+		if (!dev->rx.isPending) {
+			err = -1;
+			break;
+		}
+
+		if (completion->err != 0) {
+			dev->rx.msg.o.err = -abs(completion->err);
+		}
+		else {
+			memcpy(dev->rx.msg.o.data, data, len);
+			dev->rx.msg.o.size = len;
+			dev->rx.msg.o.err = len;
+		}
+
+		err = msgRespond(usbwlan_common.msgport, &dev->rx.msg, dev->rx.rid);
+		dev->rx.isPending = false;
+	} while (0);
+	mutexUnlock(dev->rx.lock);
+
+	usbwlan_put(dev);
+
+	return err;
+}
+
+
+static void usbwlan_close(usbwlan_dev_t *dev)
+{
+	mutexLock(dev->rx.lock);
+	dev->rx.isPending = false;
+	mutexUnlock(dev->rx.lock);
+
+	dev->flags = 0;
+	dev->clientpid = 0;
+}
+
+
+static int usbwlan_open(usbwlan_dev_t *dev, int flags, pid_t pid)
+{
+	if (dev->clientpid != 0 || dev->flags != 0) {
```

Comment:
Common variables access should be protected by mutex. Otherwise this can fail for multiple message threads.